api/apiclient_test.go

```diff
@@ -153,14 +154,91 @@ func (s *apiclientSuite) TestServerRoot(c *gc.C) {
 }
 
 func (s *apiclientSuite) TestDialWebsocketStopped(c *gc.C) {
+	f := (*api.NewWebsocketDialer)(nil, api.DialOpts{})
 	stopped := make(chan struct{})
-	f := api.NewWebsocketDialer(nil, api.DialOpts{})
 	close(stopped)
 	result, err := f(stopped)
 	c.Assert(err, gc.Equals, parallel.ErrStopped)
 	c.Assert(result, gc.IsNil)
 }
 
+func (s *apiclientSuite) TestOpenWithSNIHostNameEmptyCert(c *gc.C) {
+	info := &api.Info{
+		Addrs:       []string{"foo.com:1234", "0.1.2.3:1234"},
+		SNIHostName: "foo.com",
+		SkipLogin:   true,
+	}
+	s.testSNIHostName(c, info, []apiDialInfo{{
+		location:   "wss://foo.com:1234/api",
+		hasRootCAs: false,
+		serverName: "foo.com",
+	}, {
+		location:   "wss://0.1.2.3:1234/api",
+		hasRootCAs: false,
+		serverName: "foo.com",
+	}})
+}
+
+func (s *apiclientSuite) TestOpenWithSNIHostNameWithCACert(c *gc.C) {
+	info := &api.Info{
+		Addrs:       []string{"foo.com:1234", "0.1.2.3:1234"},
+		SNIHostName: "foo.com",
+		SkipLogin:   true,
+		CACert:      jtesting.CACert,
+	}
+	s.testSNIHostName(c, info, []apiDialInfo{{
+		location:   "wss://foo.com:1234/api",
+		hasRootCAs: true,
+		serverName: "juju-apiserver",
+	}, {
+		location:   "wss://0.1.2.3:1234/api",
+		hasRootCAs: true,
+		serverName: "juju-apiserver",
+	}})
+}
+
+type apiDialInfo struct {
+	location   string
+	hasRootCAs bool
+	serverName string
+}
+
+// testSNIHostName tests that when the API is dialed with the given info,
+// api.newWebsocketDialer is called with the expected information
+// (one element for each call to newWebsocketDialer)
+func (s *apiclientSuite) testSNIHostName(c *gc.C, info *api.Info, expectDials []apiDialInfo) {
+	dialed := make(chan *websocket.Config)
+	s.PatchValue(api.NewWebsocketDialer, func(cfg *websocket.Config, opts api.DialOpts) func(<-chan struct{}) (io.Closer, error) {
+		return func(<-chan struct{}) (io.Closer, error) {
+			dialed <- cfg
+			return nil, errors.New("nope")
+		}
+	})
+	done := make(chan struct{})
+	go func() {
+		defer close(done)
+		conn, err := api.Open(info, api.DialOpts{})
+		c.Check(conn, gc.Equals, nil)
+		c.Check(err, gc.ErrorMatches, `nope`)
+	}()
+	for _, expect := range expectDials {
```

Comment:
After changing the test and making it work, I'm no longer so sure it's better.
I think the new version is more brittle - it presupposes more about the internals
of the code being tested and will fail in a not-very-nice way (panic or hang forwever) if that
doesn't behave as expected.

I think the new version imposes somewhat more cognitive load, as the reader needs
to figure out why it's OK to close the channel.

The old version is longer, it's true, but has a more obvious failure mode - every
receive is guarded by a select so it's immediately clear that the test can't hang
forever, and there's no implicit ordering constraint between the call to the dial
function and the return from api.Open.

I think I'd prefer to stick with the original if that's OK.